### ISSUE TITLE
Randomness should be consistent across devices with use_deterministic_algorithms

### ISSUE THREAD

**Issue by antoche** [pytorch/pytorch#84234]
### 🚀 The feature, motivation and pitch

Being able to generate reproducible results is a key part of the research process, as as a research platform, pytorch should support this goal. It currently does this by providing a "deterministic " option as described in https://pytorch.org/docs/stable/notes/randomness.html. However, even with that option enabled, random number generation is inconsistent across GPUs.

I think it would be reasonable to expect to get consistent random numbers when using the same seed, PyTorch version, and CUDA version even when running on a different device (whether CPU or physical GPU). From what I understand of random number generation, hardware differences such as processor counts or FPU precision don't need to play a role in RNG. And if they do, I'm guessing a slower, deterministic pathway could be chosen when `torch.use_deterministic_algorithms()` is used.

### Alternatives

The current alternatives I know of are:

1. Always use CPU (which seems to be consistent, although I am not sure that is guaranteed) as the target for random number generation 
2. Use a different, deterministic generator (e.g., numpy)
3. Use the same hardware devices everywhere

Options 1 and 2 require all researchers to know about this issue and write their code accordingly. Unfortunately, people are not aware of it. When they write code with determinism in mind, they typically allow for a torch generator to be provided as an API argument, assuming this will guarantee determinism. This is unfortunately not the case (and is further compounded by #62451). 

Option 3 is unfortunately unrealistic and can't be sustained over time as hardware becomes obsolete.

### Additional context

Related issues: #62467 #79496 #62451

**Comment by ezyang** [comment]
Unfortunately the deterministic option is specifically for determinism under the same configuration.

It is reasonable to want a portable CUDA RNG that is consistent over devices, but unfortunately we don't have one right now.

**Comment by antoche** [comment]
As an example, here is a piece of code that I am currently using in various projects and which addresses the issues above, showing that pytorch might already contain most of the solutions to this problem:

```
def randn(size, generator=None, device=None, **kwargs):
    """
    Wrapper around torch.randn providing proper reproducibility.

    Generation is done on the given generator's device, then moved to the
    given ``device``.

    Args:
        size: tensor size
        generator (torch.Generator): RNG generator
        device (torch.device): Target device for the resulting tensor
    """
    # FIXME: generator RNG device is ignored and needs to be passed to torch.randn (torch issue #62451)
    rng_device = generator.device if generator is not None else device
    image = torch.randn(size, generator=generator, device=rng_device, **kwargs)
    image = image.to(device=device)
    return image


def randn_like(tensor, generator=None, **kwargs):
    return randn(tensor.shape, layout=tensor.layout, generator=generator, device=tensor.device, **kwargs)
```

Switching to the `randn` and `randn_like` functions above means:
- The RNG will always happen on the `Generator`'s `device`, regardless of which `device` is passed to `randn` (addressing #62451)
- The resulting tensor will still be on the requested `device`, therefore the `randn` contract is still respected
- For full determinism, the user can choose to pass a `Generator` targeting the `"cpu"` device (which currently seems to be deterministic) with a deterministic seed. 

One could go a step further and force `rng_device` to always be `"cpu"` if `torch.use_deterministic_algorithms` is set.

In the end, I'm guessing a deterministic solution which generates on the target device is preferable for performance reasons, rather than constantly going from CPU to GPU. So this example may not be the solution we're after. But it feels like a fix is tantalizingly within reach, and this issue has wide-ranging effects on all pytorch-based libraries when it comes to reliability.

**Comment by kurtamohler** [comment]
We could potentially add a `rng_device` flag to the `torch.use_deterministic_algorithms` state. So for instance, if someone does `torch.use_deterministic_algorithms(True, rng_device='cuda:0')`, then 'cuda:0' will be used for all rng. If `torch.rand(10, device='cpu')` is called, it will actually be generated on 'cuda:0', but then copied to 'cpu'. Likewise, any other operation that uses rng internally would follow suit.

However, there is a broader problem. Many of the other operations in PyTorch give different results on CPU and CUDA, and there isn't really a way to prevent that--different hardware and different device-specific software are just bound to give different results sometimes. Since we would still have this inter-device nondeterminism for compute operations, is it actually useful to force just the rng part to have inter-device determinism?

**Comment by mruberry** [comment]
> We could potentially add a `rng_device` flag to the `torch.use_deterministic_algorithms` state. So for instance, if someone does `torch.use_deterministic_algorithms(True, rng_device='cpu')`, then 'cpu' will be used for all rng. If `torch.rand(10, device='cuda')` is called, it will actually be generated on 'cpu', but then copied to 'cuda'. Likewise, any other operation that uses rng internally would follow suit. So it's sort of like a "faked" or "apparent" inter-device deterministic rng.
> 
> However, there is a broader problem. Many of the other operations in PyTorch give different results on different cpus and different gpus, and there isn't really a way to prevent that--different hardware and different device-specific software are just bound to give different results sometimes. Even on just the CPU, a vectorized operation sometimes gives different results than a non-vectorized operation, for instance. Since we would still have lots of sources of inter-device nondeterminism for compute operations, is it actually useful to force just the rng part to have apparent inter-device determinism?

In fairness to this request, however, differences in computation across devices are typically numerical accuracy issues, not completely different semantics, as they can be for randomness. To meet this request I think we'd have to define the semantics of our random number generation, possibly using a technique like JAX does.

**Comment by mruberry** [comment]
cc @t-vi who expressed some interest in this problem

**Comment by ezyang** [comment]
We have talked about implementing philox on cpu so that they line up. Stillworth doing probably

**Comment by t-vi** [comment]
Thanks @mruberry  for tagging me.

I have a lengthy write-up in progress.
TLDR you would likely need
- a common RNG (e.g. implementing philox on CPU),
- a common method for "which random number goes where" when generating an array of random numbers,
- a common method for mapping from RNG output (say uniform random numbers) to those drawn from a distribution. Traditionally, this has been distribution-specific (e.g. Box-Muller) for computational efficiency, but one might just use inverse distributions for everything to exclude variation.

I would be very glad to work on these.

**Comment by antoche** [comment]
> Since we would still have lots of sources of inter-device nondeterminism for compute operations, is it actually useful to force just the rng part to have apparent inter-device determinism?

As @mruberry mentioned, differences between RNG and other operations are quite a different category. As far as I can see, most hardware differences come as minute numerical precision issues, which can often be tolerated. But RNG differences give completely different results entirely.

Production software need to be reliable and deterministic, as well as performant. Ideally, as a user and author of production code, I would rather avoid having to enable `torch.use_deterministic_algorithms` for the sole purpose of getting a deterministic RNG when it comes with performance degradation in other areas. For similar performance reasons, it would be ideal to avoid the transfer of tensors across devices just to get deterministic RNG behaviour.

**Comment by ezyang** [comment]
Well we should offer cpu philox regardless. A non performant version shouldn't be that hard to implement, though it is a little tricky deciding how to expose it api wise

**Comment by iamgahr** [comment]
The mapping question raised above ("whether we actually do have a purposeful determinism in the mapping of array index to sequence number/counter") has a concrete answer, measured: there is none across GPU models, and the boundary is exactly one wave of threads.

In `aten/src/ATen/native/cuda/DistributionTemplates.h` the Philox subsequence is the *thread* index, and the launch grid is capped at `multiProcessorCount × (maxThreadsPerMultiProcessor / 256)` blocks (`calc_execution_policy`). For `n ≤ 256 × SMs × blocks/SM` every card produces the same seeded stream; above it, each SM count is its own stream — which is why #87992's 8 × 77 × 1024 tensor differs across GPUs and small tensors do not, and why #194676 found the T4/L4 boundary at exactly 40,960 = 256 × 40 × 4.

A CPU implementation of that mapping (Philox4x32-10, cuRAND's `curand_init(seed, subsequence, offset)` layout, `_curand_uniform`) reproduces `torch.rand(n, device="cuda", generator=...)` **bit for bit** at 2^12 … 2^22 elements on an A10G, an L40, an L40S and an H100 PCIe, given only each card's SM count and threads per SM. So a seeded device stream is portable in principle — as a transcription from (SM count, threads/SM) — and an element-indexed counter mapping would make it portable by construction. Details, the script and the measurements: #196613.
